main.go

```diff
@@ -0,0 +1,323 @@
+
+	//encode newTask as json, and return it to user.
+	if err := json.NewEncoder(w).Encode(newTask); err != nil {
+		panic(err)
+	}
+}
+
+func UpdateTask(w http.ResponseWriter, r *http.Request) {
+	//note: r.FormValue searches for key in POST data fields, then PUT data fields
+	//2 types of POST submissions: application/x-www-form-urlencoded AND multipart/form-data.
+	// need to understand both. Generally speaking, urlencoded takes up extra space so is for normal post requests. multipart form-data does not increase space usage by a lot so is for uploading files
+	//http://stackoverflow.com/a/4073451/4710047
+
+	log.Printf("Update Task")
+	//create a variable that has the parameters sent to the api in the url
+	// /task/<id>/<x>/  will lead to variables id and x in vars
+	vars := mux.Vars(r)
+
+	//get task id
+	taskId, err := strconv.Atoi(vars["id"])
+	if err != nil {
+		log.Fatal(err)
+	}
+	//get new name from POST data form
+	newName := r.PostFormValue("name")
+
+	//start the transaction (do not have to close transaction, but do have to commit them)
+	tx, err := db.Begin()
+	if err != nil {
+		log.Fatal(err)
+	}
+	//update query statement.
+	stmt, err := tx.Prepare("UPDATE Task SET  name=? WHERE id=?")
+	if err != nil {
+		log.Fatal(err)
+	}
+	//close statement
+	defer stmt.Close()
+	//check for errors. response does not give you the row that had the update so _
+	_, err = stmt.Exec(newName, taskId)
+	if err != nil {
+		log.Fatal(err)
+	}
+	//commit the transaction
+	tx.Commit()
+
+	//create model for updated task
+	updatedTask := models.Task{Id: int(taskId), Name: newName}
+
+	//encode task as json and return to user
+	if err := json.NewEncoder(w).Encode(updatedTask); err != nil {
+		panic(err)
+	}
+}
+
+//view a single task
+//input: id of task as url parameter
+//output: json serialization of task with input id
+func ViewTask(w http.ResponseWriter, r *http.Request) {
+
+	//note: r.FormValue searches for key in GET queries, then POST data fields, then PUT data fields
+	log.Printf("View Task")
+	//create vars variable to access the url parameters
+	vars := mux.Vars(r)
+
+	//get id from url params
+	taskId, err := strconv.Atoi(vars["id"])
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	//query
+	rows, err := db.Query("select id, name from Task where id=?", taskId)
+	if err != nil {
+		log.Fatal(err)
+	}
+	//close rows
+	defer rows.Close()
+
+	var id int
+	var name string
+	//make row cursor point to first and only row
+	rows.Next()
+	//fill in id and name variables
+	rows.Scan(&id, &name)
+	//create task model with given id and name
+	curTask := models.Task{Id: id, Name: name}
+	//serialize task and then output it to user as json
+	if err := json.NewEncoder(w).Encode(curTask); err != nil {
+		panic(err)
+	}
+}
+
+//return a list of all tasks
+func ViewAllTasks(w http.ResponseWriter, r *http.Request) {
+	log.Printf("View All Tasks")
+	//query for tasks
+	rows, err := db.Query("select id, name from Task")
```

Comment:
can shorten this code plus add in the error handling row wise instead of columnwise by doing: 

for s, err := c.Query(sql); err == nil; err = s.Next() {
	var rowid int64
	s.Scan(&rowid, row)     // Assigns 1st column to rowid, the rest to row
	fmt.Println(rowid, row) // Prints "1 map[a:1 b:demo c:<nil>]"
}